client/emv/emvcore.c

```diff
@@ -23,11 +94,370 @@ void TLVPrintFromBuffer(uint8_t *data, int datalen) {
 	struct tlvdb *t = NULL;
 	t = tlvdb_parse_multi(data, datalen);
 	if (t) {
-		PrintAndLog("TLV decoded:");
+		PrintAndLog("-------------------- TLV decoded --------------------");
 		
 		tlvdb_visit(t, print_cb, NULL, 0);
 		tlvdb_free(t);
 	} else {
 		PrintAndLog("TLV ERROR: Can't parse response as TLV tree.");
 	}
 }
+
+void TLVPrintFromTLV(struct tlvdb *tlv) {
+	if (!tlv) 
+		return;
+	
+	tlvdb_visit(tlv, print_cb, NULL, 0);
+}
+
+void TLVPrintAIDlistFromSelectTLV(struct tlvdb *tlv) {
+	PrintAndLog("|------------------|--------|-------------------------|");
+	PrintAndLog("|    AID           |Priority| Name                    |");
+	PrintAndLog("|------------------|--------|-------------------------|");
+
+	struct tlvdb *ttmp = tlvdb_find(tlv, 0x6f);
+	if (!ttmp)
+		PrintAndLog("|                         none                        |");
+		
+	while (ttmp) {
+		const struct tlv *tgAID = tlvdb_get_inchild(ttmp, 0x84, NULL);
+		const struct tlv *tgName = tlvdb_get_inchild(ttmp, 0x50, NULL);
+		const struct tlv *tgPrio = tlvdb_get_inchild(ttmp, 0x87, NULL);
+		if (!tgAID)
+			break;
+		PrintAndLog("|%s|   %s  |%s|", 
+			sprint_hex_inrow_ex(tgAID->value, tgAID->len, 18), 
+			(tgPrio) ? sprint_hex(tgPrio->value, 1) : "   ", 
+			(tgName) ? sprint_ascii_ex(tgName->value, tgName->len, 25) : "                         ");
+		
+		ttmp = tlvdb_find_next(ttmp, 0x6f);
+	}
+
+	PrintAndLog("|------------------|--------|-------------------------|");
+}
+
+
+int EMVSelect(bool ActivateField, bool LeaveFieldON, uint8_t *AID, size_t AIDLen, uint8_t *Result, size_t MaxResultLen, size_t *ResultLen, uint16_t *sw, struct tlvdb *tlv) {
+	uint8_t data[260] = {0};
+	*ResultLen = 0;
+	if (sw)	*sw = 0;
+	uint16_t isw = 0;
+	
+	// select APDU
+	data[0] = 0x00;
+	data[1] = 0xA4;
+	data[2] = 0x04;
+	data[3] = 0x00;
+	data[4] = AIDLen;
+	memcpy(&data[5], AID, AIDLen);
+	
+	if (ActivateField)
+		DropField();
+	
+	if (APDULogging)
+		PrintAndLog(">>>> %s", sprint_hex(data, AIDLen + 6));
+
+	int res = ExchangeAPDU14a(data, AIDLen + 6, ActivateField, LeaveFieldON, Result, (int *)ResultLen);
```

Comment:
`MaxResultLen` is not forwarded to `ExchangeAPDU14a()`. In fact it is not used at all?